File: src/main/java/net/sf/rails/game/state/SetChange.java
package net.sf.rails.game.state;

/**
 * Change associated with HashSetState
 */
public final class SetChange<E> extends Change {
    private final SetState<E> state;
    private final E element;
    private final boolean addToSet;

    /**
     * Add/Remove element to/from the set
     */
    SetChange(SetState<E> state, E element, boolean addToSet){
        this.state = state;
        this.element = element;
        this.addToSet = addToSet;
        super.init(state);
    }

    @Override 
    void execute() {
        state.change(element, addToSet);
    }

    @Override 
    void undo() {
        state.change(element, !addToSet);
    }
    
    @Override
    public SetState<E> getState() {
        return state;
    }
    
    public boolean isAddToSet() {
        return addToSet;
    }
    
    public E getElement() {
        return element;
    }
    
    @Override
    public String toString() {
        if (addToSet) {
            return "Change for " + state + ": Add " + element;
        } else {
            return "Change for " + state + ": Remove " + element;
        }
    }

}

File: dotnet/RailsLib.Net/Net/Game/State/SetChange.cs
using System;


namespace GameLib.Net.Game.State
{
    public sealed class SetChange<T> : Change
    {
        private readonly SetState<T> state;
        private readonly T element;
        private readonly bool addToSet;

        /**
         * Add/Remove element to/from the set
         */
        public SetChange(SetState<T> state, T element, bool addToSet)
        {
            this.state = state;
            this.element = element;
            this.addToSet = addToSet;
            base.Init(state);
        }

        public override void Execute()
        {
            state.Change(element, addToSet);
        }

        public override void Undo()
        {
            state.Change(element, !addToSet);
        }

        override public GameState GameState
        {
            get
            {
                return state;
            }
        }

        public bool IsAddToSet
        {
            get
            {
                return addToSet;
            }
        }

        public T Element
        {
            get
            {
                return element;
            }
        }

        override public string ToString()
        {
            if (addToSet)
            {
                return "Change for " + state + ": Add " + element;
            }
            else
            {
                return "Change for " + state + ": Remove " + element;
            }
        }

    }
}
